### PR TITLE
Gradient enablement bug fix

### PR DESCRIPTION
**Pull request by ksivaman**
The use of `self.training` attribute is meant for figuring out whether the model is an `eval` mode for layers like `dropout`/`BatchNorm` etc. that have different behavior during training and evaluation. Gradients can still be computed in eval mode but we currently disable it when using this flag instead of checking whether or not grad is enabled.

The example use case below is what this PR fixes:
```
from transformer_engine import pytorch as te

inp = torch.randn(2, 2, device="cuda", requires_grad=True)
m = te.Linear(2, 2).cuda().eval()

out = m(inp)
loss = out.mean()

torch.autograd.grad(loss, inp)
```

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

### CODE REVIEW

**Comment by netaz** [comment]
Looks good.

**Review by ptrendx — APPROVED** [review]
LGTM
